executor/analyze_test.go

```diff
@@ -1159,3 +1159,370 @@ func (s *testSuite10) TestAdjustSampleRateNote(c *C) {
+	c.Assert(err, IsNil)
+	tableInfo = table.Meta()
+	c.Assert(tableInfo.StatsOptions, NotNil)
+	c.Assert(tableInfo.StatsOptions.TopN, Equals, uint64(4))
+
+	// alter table with default
+	tk.MustExec("alter table t STATS_OPTIONS=default")
+	is = tk.Se.(sessionctx.Context).GetInfoSchema().(infoschema.InfoSchema) // refresh infoschema
+	table, err = is.TableByName(model.NewCIStr("test"), model.NewCIStr("t"))
+	c.Assert(err, IsNil)
+	tableInfo = table.Meta()
+	c.Assert(tableInfo.StatsOptions, NotNil)
+	c.Assert(tableInfo.StatsOptions.TopN, Equals, uint64(0))
+	// use global settings as default
+	tk.MustExec("insert into t values (22)")
+	c.Assert(h.DumpStatsDeltaToKV(handle.DumpAll), IsNil)
+	c.Assert(h.Update(is), IsNil)
+	h.HandleAutoAnalyze(is)
+	tbl = h.GetTableStats(tableInfo)
+	col = tbl.Columns[1]
+	c.Assert(len(col.TopN.TopN), Equals, int(col.Count))
+	tk.MustExec("analyze table t")
+	tbl = h.GetTableStats(tableInfo)
+	col = tbl.Columns[1]
+	c.Assert(len(col.TopN.TopN), Equals, int(col.Count))
+
+	// alter table with value exceeds limit
+	tk.MustExec("alter table t STATS_OPTIONS='{\"TOPN\":\"200000\"}'")
+	// refresh infoschema
+	is = tk.Se.(sessionctx.Context).GetInfoSchema().(infoschema.InfoSchema)
+	table, err = is.TableByName(model.NewCIStr("test"), model.NewCIStr("t"))
+	c.Assert(err, IsNil)
+	tableInfo = table.Meta()
+	c.Assert(tableInfo.StatsOptions, NotNil)
+	c.Assert(tableInfo.StatsOptions.TopN, Equals, uint64(200000))
+	c.Assert(tableInfo.StatsOptions.Buckets, Equals, uint64(0)) // buckets-num is overwritten to 0
+	execErr := tk.ExecToErr("analyze table t")                  // error occurs at planning since limit could change
+	c.Assert(execErr, NotNil)
+}
+
+func (s *testSuite10) TestAnalyzeOptionsForPartitionTable(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("set @@tidb_analyze_version=2")
+	createTable := `CREATE TABLE t (a int, b int, c varchar(10), primary key(a), index idx(b)) STATS_TOPN=1
+PARTITION BY RANGE ( a ) (
+		PARTITION p0 VALUES LESS THAN (6),
+		PARTITION p1 VALUES LESS THAN (11),
+		PARTITION p2 VALUES LESS THAN (16),
+		PARTITION p3 VALUES LESS THAN (21)
+)`
+	tk.MustExec(createTable)
+	for i := 1; i < 20; i++ {
+		tk.MustExec(fmt.Sprintf(`insert into t values (%d, %d, "hello")`, i, i))
+	}
+
+	handle.AutoAnalyzeMinCnt = 0
+	tk.MustExec("set global tidb_auto_analyze_ratio = 0.01")
+	is := tk.Se.(sessionctx.Context).GetInfoSchema().(infoschema.InfoSchema)
+	table, err := is.TableByName(model.NewCIStr("test"), model.NewCIStr("t"))
+	c.Assert(err, IsNil)
+	tableInfo := table.Meta()
+	h := s.domain.StatsHandle()
+	pi := tableInfo.GetPartitionInfo()
+	c.Assert(pi, NotNil)
+
+	// manual analyze first
+	tk.MustExec("analyze table t")
+	for _, def := range pi.Definitions {
+		partition := h.GetPartitionStats(table.Meta(), def.ID)
+		pcol := partition.Columns[1]
+		c.Assert(len(pcol.TopN.TopN), Equals, 1)
+	}
```

Comment:
Maybe we can set `set @@tidb_partition_prune_mode='dynamic'` and check the length of TopN for the global table stats.